### PR TITLE
feat: Adds the ECharts Bubble chart

### PR DESCRIPTION
**Pull request by mayurnewase**
### SUMMARY
Adds the ECharts Bubble chart.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
1. Chart and data control panel:
Screenshot from 2022-11-23 10-52-20

2. Customize control panel
Screenshot from 2022-11-27 16-18-56

3. Tooltip:
Screenshot from 2022-11-23 11-11-23

### TESTING INSTRUCTIONS
1. create d3 bubble chart
2. merge commits from this PR and open existing chart should work fine as before.
4. test all controls in `data` and `customize` sections.

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [x] Changes UI
- [x] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [x] Migration is atomic, supports rollback & is backwards-compatible
  - [x] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset-frontend/plugins/plugin-chart-echarts/src/Bubble/transformProps.ts
Comment (mayurnewase) [inline superset-frontend/plugins/plugin-chart-echarts/src/Bubble/transformProps.ts:57]: didn't use echart's sanitize function, it removes line breaks.

File: superset-frontend/.storybook/main.js
Comment (mayurnewase) [inline superset-frontend/.storybook/main.js:27]: bycatch: should include stories for plugins when storybook server is run.

File: superset-frontend/plugins/plugin-chart-echarts/src/Bubble/index.ts
Comment (mayurnewase) [inline superset-frontend/plugins/plugin-chart-echarts/src/Bubble/index.ts:42]: will revert to original `Bubble Chart` in next commit.

File: superset-frontend/plugins/plugin-chart-echarts/src/Bubble/transformProps.ts
Comment (mayurnewase) [inline superset-frontend/plugins/plugin-chart-echarts/src/Bubble/transformProps.ts:43]: `* 2` to have same size options as previous charts in the control panel to avoid db migration.

File: superset-frontend/plugins/plugin-chart-echarts/src/Bubble/transformProps.ts
Comment (villebro) [inline superset-frontend/plugins/plugin-chart-echarts/src/Bubble/transformProps.ts:97]: This will only work for adhoc metrics, use `getMetricLabel` from `@superset-ui/core` to make it also support saved metrics: 
```suggestion
  const xAxisLabel: string = getMetricLabel(x);
  const yAxisLabel: string = getMetricLabel(y);
  const sizeLabel: string = getMetricLabel(size);
```

File: superset-frontend/plugins/plugin-chart-echarts/src/Bubble/transformProps.ts
Comment (villebro) [inline superset-frontend/plugins/plugin-chart-echarts/src/Bubble/transformProps.ts:109]: missed this one:
```suggestion
          datum[sizeLabel],
```

File: superset-frontend/plugins/plugin-chart-echarts/src/Bubble/transformProps.ts
Comment (villebro) [inline superset-frontend/plugins/plugin-chart-echarts/src/Bubble/transformProps.ts:171]: `defaultTooltip` seems to have been a leftover from a recent fix (my bad; I should have removed it in #22218). Please use `getDefaultTooltip` from `utils/tooltip.ts` (this ensures proper tooltip positioning) and remove `defaultTooltip` in `defaults.ts`:
```suggestion
      ...getDefaultTooltip(refs),
```

File: superset-frontend/plugins/plugin-chart-echarts/src/Bubble/transformProps.ts
Comment (villebro) [inline superset-frontend/plugins/plugin-chart-echarts/src/Bubble/transformProps.ts:60]: We should probably use the formatters here for x and y values. In some other plugins I believe we have separate formatters for the axes and the tooltip - I think that could be appropriate here. Also, should we perhaps have a formatter for the bubble size, too?

**Review by villebro — COMMENTED** [review]
First pass comments, looks great! One thing that I noticed is that the NVD3 plugin assumes truncation on the y-axis if the bounds aren't set. While I don't necessarily like the idea of defaulting to truncation (actually, I strongly oppose it), I was just wondering if we should do a migration that enables truncation for existing charts if the bounds are undefined?

File: superset-frontend/plugins/plugin-chart-echarts/src/Bubble/transformProps.ts
Comment (mayurnewase) [reply superset-frontend/plugins/plugin-chart-echarts/src/Bubble/transformProps.ts:60]: should we have single formatter for all 3 values?

Comment (mayurnewase) [reply superset-frontend/plugins/plugin-chart-echarts/src/Bubble/transformProps.ts:60]: having 3 seaparate seems bit overkill.

Comment (mayurnewase) [reply superset-frontend/plugins/plugin-chart-echarts/src/Bubble/transformProps.ts]: added 1, let me know if we need 3 separate controls.

Comment (villebro) [reply superset-frontend/plugins/plugin-chart-echarts/src/Bubble/transformProps.ts]: Looking at the example chart in the World Bank Dashboard, the size is in billions, while t x-axis and y-axis are essentially percentages:
image

I think x and y axes will usually be in the same scale, but since we already have separate formatters for them and no separate tooltip formatter, I would maybe just propose adding a size formatter, and then using the same formatters in the tooltip for now (later on if people really want to use separate formatters on the tooltip we can add that, but let's not go for 6 formatters quite yet 😬).

Comment (mayurnewase) [reply superset-frontend/plugins/plugin-chart-echarts/src/Bubble/transformProps.ts]: added.

**Comment by mayurnewase** [comment]
> First pass comments, looks great! One thing that I noticed is that the NVD3 plugin assumes truncation on the y-axis if the bounds aren't set. While I don't necessarily like the idea of defaulting to truncation (actually, I strongly oppose it), I was just wondering if we should do a migration that enables truncation for existing charts if the bounds are undefined?

agree, added migration.

File: superset-frontend/plugins/plugin-chart-echarts/src/Bubble/transformProps.ts
Comment (villebro) [inline superset-frontend/plugins/plugin-chart-echarts/src/Bubble/transformProps.ts:200]: we need to pass refs here to the component:
```suggestion
    refs,
    height,
```

File: superset-frontend/plugins/plugin-chart-echarts/src/Bubble/transformProps.ts
Comment (villebro) [inline superset-frontend/plugins/plugin-chart-echarts/src/Bubble/transformProps.ts:179]: Let's make the legend customizable like in the other charts, both so it can be disabled/enabled and also changeing the orientation etc (there should be controls and utils that can be readily reused)

**Review by villebro — COMMENTED** [review]
A few more things that came to mind. Also, one major issue: the chart is heavily padded:
image
I quickly verified that by adding `left: 0, right: 0` to the grid seemed to widen it to the full width, although it went a few pixels over. So I think we need to reuse the `getPadding` util from the Timeseries chart to get the grid right.

**Review by zhaoyongjie — COMMENTED** [review]
I just saw the original bubble chart(NVD3 version), and the `limit control` have an inconsistent setting. IMO, we should fix/improve that in the __new version__ of Bubble Chart.

1. The `Series Limit` refers to a limit by series rather than a `row limit`.
2. No chance to set a `order by` clause in the SQL

image

File: superset-frontend/plugins/plugin-chart-echarts/src/Bubble/constants.ts
Comment (zhaoyongjie) [inline superset-frontend/plugins/plugin-chart-echarts/src/Bubble/constants.ts:22]: Could remove the `ts-ignore` in the new plugins?

Comment (mayurnewase) [reply superset-frontend/plugins/plugin-chart-echarts/src/Bubble/constants.ts:22]: its because `SqlaFormData` has many values which are not set default like viz_type, datasource etc.

**Comment by mayurnewase** [comment]
> I just saw the original bubble chart(NVD3 version), and the `limit control` have an inconsistent setting. IMO, we should fix/improve that in the **new version** of Bubble Chart.
> 
> 1. The `Series Limit` refers to a limit by series rather than a `row limit`.
> 2. No chance to set a `order by` clause in the SQL
> 
> image

here series limit made sense as its inline with echart's options.(every bubble is separate series)
and description of `dimension` in control panel had series in it to group them by color.

File: superset-frontend/plugins/plugin-chart-echarts/src/Bubble/constants.ts
Comment (mayurnewase) [reply superset-frontend/plugins/plugin-chart-echarts/src/Bubble/constants.ts:22]: should make them optional?

Comment (mayurnewase) [reply superset-frontend/plugins/plugin-chart-echarts/src/Bubble/constants.ts:22]: it will need modification in core, which again needs 100% coverage which will increase the PR size.

**Comment by mayurnewase** [comment]
> @mayurnewase the `series limit` will send a subquery instead a simple query, the time-series chart should be referenced.

I see, let me check.

**Comment by villebro** [comment]
FYI @mayurnewase @zhaoyongjie , slightly related, I'm opening a PR today that will clean up deprecated limit control logic from superset-ui/core so we can remove those annoying deprecation warnings on chart data requests

**Comment by mayurnewase** [comment]
updated series limit to row limit, and added order control.

@zhaoyongjie the `normalizeOrderBy` in core doesn't support this control panel yet and I didn't want to change anything in core in this PR so added order support in the buildQuery itself.
Either this or need to something more hacky like below

`orderby: normalizeOrderBy(
      {
        ...baseQueryObject,
        orderby: [
          [...baseQueryObject.orderby, !baseQueryObject.order_desc]
        ]
      }
    ).orderby`

what do you think?

**Comment by rusackas** [comment]
Just seeing this and catching up on the thread - this is great, thanks!

I haven't looked at the code yet, but for myself or any other reviewer, I'm hopeful that this (or a followup) includes the latest best practices:
• Generic X Axis
• Standardized controls (normalizing/denormalizing for fast-viz switching)
• Drill to Detail support

Basically, trying to keep all the ECharts plugins current with all the first-class bells and whistles. Happy to provide support/references for any of the above if needed.

Again, thanks, this is exciting!

**Comment by mayurnewase** [comment]
> Just seeing this and catching up on the thread - this is great, thanks!
> 
> I haven't looked at the code yet, but for myself or any other reviewer, I'm hopeful that this (or a followup) includes the latest best practices: • Generic X Axis • Standardized controls (normalizing/denormalizing for fast-viz switching) • Drill to Detail support
> 
> Basically, trying to keep all the ECharts plugins current with all the first-class bells and whistles. Happy to provide support/references for any of the above if needed.
> 
> Again, thanks, this is exciting!

@rusackas 
1. x and y axes are both metrics
2. controls are standard but for fast switching don't think there is any other chart with x and y and 1 more metric for bubble size, so not sure how I can test the switch.
3. I will add cross filter and drill to detail as soon as this PR gets merged, because it may need other modifications and new tests.

possible to get 1 more review?

**Comment by rusackas** [comment]
Is there any intent to support temporal or categorical X-axis on this? 

Also, the ephemeral environment is up and running if any of the earlier reviewers wouldn't mind giving this another pass.

**Comment by rusackas** [comment]
There are a few things from the old chart that aren't supported in this new one
• Display of X bounds and Wy bounds (the extent numbers on the axes)
• Auto-rotation of axis labels
• Axis limit inputs (love 'em or hate 'em)
• Series limit (this one seems like it would be important)
• Max Bubble Size

While I think _most_ people would be fine with _most_ of the above, it might be a bit contentious to call it a drop-in replacement without complete feature parity. We've learned this lesson before and been stuck with difficult reverts. I see a few options:

1) Go for complete feature parity (which we might not want to!)
2) Post a DISCUSS thread asking for lazy consensus on the dev list, to see if others are OK with this change, and proceed accordingly
3) Add this as a new chart, mark the current one as legacy, and live with both in the product until we can make a breaking change (3.0)

I'm thinking  option 2 might be prudent. I LOVE this PR and would love to see any NVD3 chart die, but I want to make sure we're not pushing any (admittedly small) breaking changes on people without at least some form of official consent.

**Comment by mayurnewase** [comment]
1. Display of X bounds and Wy bounds (the extent numbers on the axes) -> here bounds are visible by default.

2. Axis limit inputs -> just like old chart this one has Y axis limit -> need to click on `Truncate Y Axis`
Screenshot from 2022-12-14 07-45-01


3. Auto rotation -> I didnt add this for consistency, as its not present in other echart charts.
4. Series Limit -> has been changed to Row Limit as per review from @zhaoyongjie above to be inline with newer updates.
5. Max bubble size -> its there in customize section.
Screenshot from 2022-12-14 07-44-19


I tested above on ephemeral deployment.

**Comment by rusackas** [comment]
@mayurnewase this appears to need a rebase. It might also need some commits added that change the old bubble chart name to include "legacy." @villebro has done some of this type of renaming and can probably warn us of any intricacies here.

**Comment by EugeneTorap** [comment]
Hi @mayurnewase! Do you have some time to finish the PR?

**Comment by mayurnewase** [comment]
@EugeneTorap unfortunately no, feel free to take over or close this PR.

**Comment by john-bodley** [comment]
cc @michael-s-molina for context who has been working on other EChart migrations.

File: superset-frontend/.storybook/main.js
Comment (michael-s-molina) [reply superset-frontend/.storybook/main.js:27]: Actually they are separate storybooks currently. Before merging them, we need to think about how to organize the components.

**Comment by michael-s-molina** [comment]
Hi @mayurnewase. Thank you for this contribution. I would like to continue the work and see if we can ship this feature. To do that I can push to your PR if you give me the GitHub permission or I can close this PR and open my own and reference you as the original author. Which method do you prefer?

**Comment by mayurnewase** [comment]
Hey thanks for taking this up, I am ok with both methods.
Added you as a collaborator in my fork of this repo, and I see ` Maintainers are allowed to edit this pull request.` in the right panel.

**Comment by michael-s-molina** [comment]
> Added you as a collaborator in my fork of this repo, and I see  Maintainers are allowed to edit this pull request. in the right panel.

Thanks @mayurnewase. I'll rebase the PR and modify its scope a little bit to avoid introducing a breaking change. I'll move the migration script to its own PR, similar to https://github.com/apache/superset/pull/23973 and also keep the legacy plugin. The new scope will be to just add the ECharts Bubble chart which will allow us to merge this PR without the need to wait for a major version release.

File: superset-frontend/plugins/plugin-chart-echarts/src/Bubble/constants.ts
Comment (michael-s-molina) [reply superset-frontend/plugins/plugin-chart-echarts/src/Bubble/constants.ts]: Resolved with `Partial`.

**Comment by michael-s-molina** [comment]
I deprecated the legacy chart and added thumbnails for the new version.

Screenshot 2023-09-20 at 13 41 49

**Review by villebro — APPROVED** [review]
I think this looks great. The only nit that comes to mind is different handling of the title and missing colors in the tooltip:

legacy:
image

new:
image

But as discussed a few months ago, I think we have some tooltip harmonization and re-styling work to do anyway, so maybe this can be deferred to that effort.

**Comment by michael-s-molina** [comment]
Thanks for the review @villebro!

> But as discussed a few months ago, I think we have some tooltip harmonization and re-styling work to do anyway, so maybe this can be deferred to that effort.

Yes, let's tackle this in a follow-up as part of the harmonization and re-styling work.

**Pull request merged**
